### DISCUSSION TITLE
Please explain the (list?) presented usig a colon :

### DISCUSSION THREAD

**Discussion by MeghansUncle2** [micropython/micropython#11481] [Using MicroPython] [answered]
I am looking through an SPI TFT MicroPython driver  and came upon the following Code (I am assuming it's a list, no?):

```py
    ROTATE = {
          0: 0x88,
         90: 0xE8,
        180: 0x48,
        270: 0x28
    }
```
It is used to validate the rotation value given.  But I dont understand the list's structure.  What are the hex values for?  They don't seem to serve any purpose.

The list is used as follows:

```py
if rotation not in self.ROTATE.keys():
            raise RuntimeError('Rotation must be 0, 90, 180 or 270.')
        else:
            self.rotation = self.ROTATE[rotation]
```
are the 0, 90, 180, and 270 values what ".keys" returns?  Or does it return the hex values, and that's what "rotation" must contain?

**Comment by robert-hh** [comment, marked as answer]
The value assigned to ROTATE is a dictionary. That consists of a sequence of key:value pairs.  
ROTATE.keys() returns the list of keys. ROTATE][key] retrieves the value of the respective key:value pair. For more information about dictionaries please consult any Python book.

**Reply by MeghansUncle2** [reply]
Awesome.  I thought it was a list.  Thank you!

**Comment by bulletmark** [comment]
Also that code can be improved. The statement `if rotation not in self.ROTATE.keys()` should just be `if rotation not in self.ROTATE` because Python (and MP) checks for existence in the dict keys by default. However, you are actually doing this check twice anyhow, once in the existence check and then again in the fetch. So you should just do both at the same time:

```python
self.rotation = self.ROTATE.get(rotation)
if self.rotation is None:
    raise RuntimeError(f'Rotation must be one of: {self.ROTATE.keys()})
```

Note also that I removed the second hard-coding of the ROTATE values which you had in the error text.

**Reply by MeghansUncle2** [reply]
Much cleaner, thanks.  I am just learning this language.  All tips welcome!

**Reply by MeghansUncle2** [reply]
@bulletmark ,

Is the RuntimeError statement you provided malformed (the **f** before **'Rotation** and no closing tic mark after **of:**  or concatenation for the derived data **{self.ROTATE.keys()}** at the end?

**Reply by bulletmark** [reply]
Sorry, that is what you get when you post code without trying it. That last line should be:
```python
raise RuntimeError(f'Rotation must be one of: {list(self.ROTATE)}')
```

**Reply by MeghansUncle2** [reply]
Got it, thanks.  And f is for f-string.  I am just learning this by reading through and tinkering.  Your help is appreciated!
